plutus-core/plutus-core/src/PlutusCore/Evaluation/Machine/CostingFun/Core.hs

```diff
@@ -39,6 +43,30 @@ import Deriving.Aeson
 import GHC.Exts
 import Language.Haskell.TH.Syntax hiding (Name, newName)
 
+-- | A class used for convenience in this module, don't export it.
+class OnMemoryUsages c a where
+    -- | Turn
+    --
+    -- > \mem1 ... memN -> f mem1 ... memN
+    --
+    -- into
+    --
+    -- > \arg1 ... argN -> f (memoryUsage arg1) ... (memoryUsage argN)
+    --
+    -- so that we don't need to repeat those 'memoryUsage' calls at every use site, which would also
+    -- require binding @arg*@ explicitly, i.e. require even more boilerplate.
```

Comment:
Yeah, this file is a bit heavy on the boilerplate.  I did wonder if we could do something with TH to reduce it a bit, but that might just obscure things.

Reply:
Yeah, I think TH generally makes things much harder to digest and it's important to keep this stuff readable. I think we could reduce the amount of boilerplate with a bit of type-level magic, but then I'm not sure if it's worth it. That `OnMemoryUsages` seems to have a great cost-to-utility ratio and I'm not sure if, say, merging all `runCostingFun*Arguments` into a single function would have a similarly favorable cost-to-utility ratio.